### DISCUSSION TITLE
Toy Keyboard integration of expanders

### DISCUSSION THREAD

**Discussion by MichaelPNolan** [marcel-licence/esp32_basic_synth#60] [Ideas]
If anyone is hacking keyboards on the cheap i've done 2 so far. So I wanted to put this info here in case its interesting to anyone.

1. I just wired into the positive side of switches on the pcb. Then if it pulled down - detect and trigger as a button. Easy. Made a midi keyboard to test my project USB using 32u4 based microproc Arduino pro micro
2. Matrix style with 37 keys - very cheap toy, like $5. It has pads - doesn't take long to figure out the columns and rows. You can wire to the columns and rows.
Then I wanted to have an ESP32 with the matrix so it's an all in one - no midi necessary. So this is where seeing a video of Marcel using GPIO expanders such as MCP23017 was a good clue. I bought them and wired up one - tested and working. 

I had already tested the matrix direct to pins of another arduino keypad.h library. I then had a bit of a palava figuring out how to use MCP23017 pins. The library of Adafruit is great but ... well it wasn't working out with this keyboard - i forked adafruit_keypad.cpp and wrote in calls for pins over 100 to call MCP23017 gpio pins via it's library. I passed a pointer to the MCP object using adafruit_mcp32017.h calls. The code worked but the keypad code wasn't working like the arduino keypad.h library had.
Then I had a whole issue about is this diodes missing and ghosting because any key of any row would trigger the whole row.

I found some libraries designed properly for doing this:  (this is what I want to document)
https://github.com/joeyoung/arduino_keypads  the keypad_mc17 (or other variations of IO expansions) are handled. It works, it also does multi-key so I can press several at once - seems to work well so far.

Now I'm integrating as many devices into this toy - a screen, pots (and another analogue port expander). I've soldered together something that is working. The only sadness was that my I2s dac board seems DOA. I have this other one that is fine with the prototype and I don't want to touch it since it's my only working one.
My goal is also to add a conventional midi out plug to this esp so I have something that drives that.

Its good to have this thown into your code so you can find out if its showing up on the SPI bus. That it's address is right - you'll need to set the address for the 

`#define I2CADDR 0x20        // address of MCP23017 chip on I2C bus`

This scan will check that you see the device and it's address is confirmed for your code. I will have another SPI device or two going on like the screen so I'll need to avoid address conflicts. I also set 4.7k ohm resistors to pull up the SPI clock and data lines - you should only do this once for the bus. This will improve communication performance.

```
void scan() {
  byte error, address;
  int nDevices;
  Serial.println("Scanning...");
  nDevices = 0;
  for(address = 1; address < 127; address++ ) {
    Wire.beginTransmission(address);
    error = Wire.endTransmission();
    if (error == 0) {
      Serial.print("I2C device found at address 0x");
      if (address<16) {
        Serial.print("0");
      }
      Serial.println(address,HEX);
      nDevices++;
    }
    else if (error==4) {
      Serial.print("Unknow error at address 0x");
      if (address<16) {
        Serial.print("0");
      }
      Serial.println(address,HEX);
    }    
  }
  if (nDevices == 0) {
    Serial.println("No I2C devices found\n");
  }
  else {
    Serial.println("done\n");
  }
  delay(5000);          
}*/
```

**Comment by MichaelPNolan** [comment]
Threw an OLED on the SPI bus ... working no problem. I call the setup and display.display(); from Task0. I call the write to screen memory from various modules. This avoids audio artifacts.

**Reply by marcel-licence** [reply]
Does display.display() transfer the data to display or just initiates to display data from the OLED's internal buffer?

**Comment by MichaelPNolan** [comment]
Well I've read that it transfers data held in ram when you call that and
that made sense with offloading to task0 to prevent audio glitches. It's so
simplistic that I don't think the screen has a frame buffer to preload but
I could be wrong. So the library creates its own framebuffer in effect.

**Comment by MichaelPNolan** [comment]
This project is moving along as a fork of basic synth. I've made my own knobs, banks of parameters physical interface in this self contained unit. 
Anyone want to see my fork 
https://github.com/MichaelPNolan/StandAloneSynth37key/tree/main/esp32_alone_synth
Ran into a lot of little electronics building issues with physically making it stable and wiring it. My projects are kind of semi-wired and semi-wires into temporary pin-connectors like you have on a test arduino where I soldered some connectors to the ESP32 dev module.
Getting it to be stable and using the right wire and lengths so I can keep opening it up and improving it without breaking some connection takes experience. I am starting to look more at the modules I was just using at how the work and learn from that.
Had some weird DAC issues when i had a floating local SCLK instead of earthing it. Sometimes I get a bit of clipping with the sinewave or the current use causes it to lose it's stability until I had the wiring done right. Started to use my on AWG22 solid core breadboard style wires cut from a spool so I could make things stable.
I like my display setup - i have 8 sectors and a routing for showing the set levels. I need a button I think to put it into edit or not edit so that I can change sets of parameters without overwriting things. 
I need to make a patch write/read or at least read out a list of numbers to save as presets.

I am milking this for all I can to develop something I have a good deal of control.
I need to make a physical 5 pin DIN midi in/out so I can slave or master out the keys.

**Reply by marcel-licence** [reply]
I will put it into the readme <derived projects>
